mining/src/cache.rs

```diff
@@ -0,0 +1,45 @@
+use consensus_core::block::BlockTemplate;
+use std::{
+    rc::Rc,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+/// CACHE_LIFETIME indicates the default duration in milliseconds after which the cached data expires.
+const DEFAULT_CACHE_LIFETIME: u64 = 1_000;
+
+pub(crate) struct BlockTemplateCache {
+    /// Time, in milliseconds, when the cache was last updated
+    last_update_time: u64,
+    block_template: Option<Rc<BlockTemplate>>,
+
+    /// Duration in milliseconds after which the cached data expires
+    cache_lifetime: u64,
+}
+
+impl BlockTemplateCache {
+    pub(crate) fn new(cache_lifetime: Option<u64>) -> Self {
+        let cache_lifetime = cache_lifetime.unwrap_or(DEFAULT_CACHE_LIFETIME);
+        Self { last_update_time: 0, block_template: None, cache_lifetime }
+    }
+
+    pub(crate) fn clear(&mut self) {
+        // This differs from golang implementation.
+        // The cache timer is reset to 0 so its lifetime is expired.
+        self.last_update_time = 0;
+        self.block_template = None;
+    }
+
+    pub(crate) fn get_immutable_cached_template(&self) -> Option<Rc<BlockTemplate>> {
+        if SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis() as u64 - self.last_update_time > self.cache_lifetime {
```

Comment:
I suggest using `i64` for `self.last_update_time` since sys time can go back if the system clock is changed so we'd like to avoid the overflow here